core/iwasm/compilation/aot_emit_gc.c

```diff
@@ -104,4 +136,1476 @@ aot_call_aot_create_func_obj(AOTCompContext *comp_ctx, AOTFuncContext *func_ctx,
+                            check_objs_succ))
+        goto fail;
+
+    /* Create if block */
+    ADD_BASIC_BLOCK(len_gt_zero, "len_gt_zero");
+    MOVE_BLOCK_AFTER_CURR(len_gt_zero);
+    /* Create inner else block */
+    ADD_BASIC_BLOCK(inner_else, "inner_else");
+    MOVE_BLOCK_AFTER(inner_else, len_le_zero);
+
+    /* Create else(end) block */
+    ADD_BASIC_BLOCK(len_le_zero, "len_le_zero");
+    MOVE_BLOCK_AFTER_CURR(len_le_zero);
+
+    BUILD_ICMP(LLVMIntSGT, len, I32_ZERO, cmp[0], "cmp_len");
+    BUILD_COND_BR(cmp[0], len_gt_zero, len_le_zero);
+
+    /* Move builder to len > 0 block */
+    SET_BUILDER_POS(len_gt_zero);
+    /* dst_offset > UINT32_MAX - len */
+    if (!(boundary = LLVMBuildAdd(comp_ctx->builder, dst_offset, len, ""))) {
+        aot_set_last_error("llvm build failed.");
+        goto fail;
+    }
+    BUILD_ICMP(LLVMIntUGT, boundary, I32_CONST(UINT32_MAX), cmp[0],
+               "boundary_check1");
+    /* dst_offset + len > wasm_array_obj_length(dst_obj) */
+    if (!aot_get_array_obj_length(comp_ctx, func_ctx, dst_obj, &array_len))
+        goto fail;
+    BUILD_ICMP(LLVMIntUGT, boundary, array_len, cmp[1], "boundary_check2");
+    /* src_offset > UINT32_MAX - len */
+    if (!(boundary = LLVMBuildAdd(comp_ctx->builder, src_offset, len, ""))) {
+        aot_set_last_error("llvm build failed.");
+        goto fail;
+    }
+    BUILD_ICMP(LLVMIntUGT, boundary, I32_CONST(UINT32_MAX), cmp[2],
+               "boundary_check3");
+    /* src_offset + len > wasm_array_obj_length(src_obj) */
+    if (!aot_get_array_obj_length(comp_ctx, func_ctx, src_obj, &array_len))
+        goto fail;
+    BUILD_ICMP(LLVMIntUGT, boundary, array_len, cmp[3], "boundary_check4");
+
+    /* logical or above 4 boundary checks */
+    for (i = 1; i < 4; ++i) {
+        if (!(cmp[0] = LLVMBuildOr(comp_ctx->builder, cmp[0], cmp[i], ""))) {
+            aot_set_last_error("llvm build failed.");
+            goto fail;
+        }
+    }
+
+    if (!aot_emit_exception(comp_ctx, func_ctx, EXCE_ARRAY_OOB, true, cmp[0],
+                            inner_else))
+        goto fail;
+
+    if (!aot_call_wasm_obj_copy(comp_ctx, func_ctx, dst_obj, dst_offset,
+                                src_obj, src_offset, len))
+        goto fail;
+
+    return true;
+fail:
+    return false;
+}
+#endif /* end of WASM_ENABLE_GC_BINARYEN != 0 */
+
+bool
+aot_compile_op_array_len(AOTCompContext *comp_ctx, AOTFuncContext *func_ctx)
+{
+    LLVMValueRef array_obj, cmp, array_len;
+    LLVMBasicBlockRef check_array_obj_succ;
+
+    POP_REF(array_obj);
+
+    ADD_BASIC_BLOCK(check_array_obj_succ, "check array obj succ");
+    MOVE_BLOCK_AFTER_CURR(check_array_obj_succ);
+
+    BUILD_ISNULL(array_obj, cmp, "cmp_array_obj");
+    if (!aot_emit_exception(comp_ctx, func_ctx, EXCE_NULL_GC_REF, true, cmp,
+                            check_array_obj_succ))
+        goto fail;
+
+    if (!aot_get_array_obj_length(comp_ctx, func_ctx, array_obj, &array_len))
```

Comment:
Don't call function, just generate LLVM IR of `array_obj->length >> WASM_ARRAY_LENGTH_SHIFT`

Reply:
This static function generates LLVM IR instead of calling a function, it's used in 3 places so I made it into a function.